app/lib/backend/storage_rewrite.rb

```diff
@@ -1,106 +1,178 @@
-      usage_limits = provider.usage_limits.includes(plan: :service)
+        logger.info "#{action} services for provider #{id}..."
+        process(provider.services)
 
-      total_count = services.size + cinstances.size + metrics.size + usage_limits.size + bought_cinstances.size
+        logger.info "#{action} buyer applications for provider #{id}..."
+        process(provider.buyer_applications)
 
-      index = 0
-      progress = -> do
-        percent = ((index + 1) / total_count.to_f) * 100.0
-        yield percent if block_given?
-        index += 1
+        logger.info "#{action} provider applications for provider #{id}..."
+        process(provider.bought_cinstances)
+
+        logger.info "#{action} metrics for provider #{id}..."
+        process(Metric.by_provider(provider))
+
+        logger.info "#{action} usage limits for provider #{id}..."
+        process(provider.usage_limits)
       end
 
-      services.find_each do |service|
-        service.update_backend_service
-        progress.call
+      private
 
-        service.service_tokens.find_each(&ServiceTokenService.method(:update_backend))
+      # Rewrite the collection
+      def process(collection)
+        rewrite(scope: collection)
       end
 
-      metrics.find_each do |metric|
-        metric.sync_backend!
-        progress.call
+      # This logger just prints out a message to STDOUT, with new line before and after.
+      # New line before is to make progress log look better
+      def logger
+        @logger ||= begin
+          log = ActiveSupport::Logger.new($stdout)
+          log.formatter = ->(_, _, _, msg) { "\n#{msg.is_a?(String) ? msg : msg.inspect}\n" }
+          log
+        end
       end
 
-      usage_limits.find_each do |usage_limit|
-        usage_limit.send(:update_backend_usage_limit)
-        progress.call
+      # All accounts eligible for backend sync
+      # Optionally include inactive (deleted and suspended)
+      def providers
+        @providers ||=
+          begin
+            providers = Account.providers_with_master
+            providers = providers.without_deleted.without_suspended unless include_inactive
+            providers
+          end
       end
+    end
 
-      update_cinstance = ->(cinstance) do
-        cinstance.send(:update_backend_application)
-        cinstance.send(:update_backend_user_key_to_application_id_mapping)
-        if cinstance.provider_account
-          cinstance.application_keys.each { |k| k.send(:update_backend_value) }
-          cinstance.referrer_filters.each { |f| f.send(:update_backend_value) }
-        end
-        progress.call
+    # Used for scheduling asynchronous jobs for later processing
+    class AsyncProcessor < Processor
+      BATCH_SIZE = 500
+
+      attr_reader :enqueuer
+
+      def initialize(**kwargs)
+        enq = kwargs.delete(:enqueuer)
+        raise ArgumentError, ':enqueuer must be present and respond to :call' unless enq.respond_to?(:call)
```

Comment:
This class is organized as if we could have multiple enqueuers. I highly doubt we would have any more though :)
What do you think about simplifying so that this initializer does not receive an enqueuer as a parameter but just enqueues to to the worker in the `#process` method?

Reply:
You mean, couple `StorageRewrite` and `BackendStorageRewriteWorker` and just hardcode `BackendStorageRewriteWorker.perform_async(batch.klass.name, batch.pluck(:id))` within the `#process` ?

Well, yeah, that would probably be more simple on the first look, but I am not sure :) I am not a fan of two-way dependencies. Currently (in the current state of the PR) we have a `StorageRewrite` module that doesn't depend on anything else, and has it's own interface, and `BackendStorageRewriteWorker` uses this module according to this interface.
So, `StorageRewrite` doesn't need to know anything about `BackendStorageRewriteWorker`'s own interface, and what it expects in `#perform`.

If we change this, it's kind of two-way, `BackendStorageRewriteWorker` calls `StorageRewrite`, and then `StorageRewrite` calls `BackendStorageRewriteWorker`, I think it might actually be more confusing. At least this is what I don't like in the Billing module, where `BillingService` calls `BillingWorker`, and then `BillingWorker` calls `BillingService`, I find it hard to find who's the chicken and who's the egg :grimacing:

Reply:
If you ask me the worker shouldn't have a method to enqueue for whole provider or for all providers. These are outside worker's area of responsibility IMO. They should be part of a service. What `StorageRewrite` actually mimics. But I didn't want to request making it a proper service or a few services :)

Reply:
Yeah, fair enough :grimacing: 
But it seems to be a very common pattern throughout our code base (which of course doesn't mean it's a good one).

![image](https://github.com/3scale/porta/assets/1270328/335ea478-2914-4615-acfd-bcedf94ec0c5)